staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -147,3 +147,21 @@ type TLSConfig struct {
 	// +optional
 	ClientCert string
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TracingConfiguration provides versioned configuration for tracing clients.
+type TracingConfiguration struct {
+	metav1.TypeMeta
+
+	// +optional
+	// URL of the collector that's running on the control-plane node.
+	// The APIServer uses the egressType ControlPlane when sending data to the collector.
+	// Defaults to the otlp grpc default, localhost:4317
```

Comment:
is a scheme allowed in this URL? if unspecified, does it default to http or https? if https, what TLS configuration is used?

Reply:
also, update doc for external config type with this info

Reply:
It must adhere to https://github.com/grpc/grpc/blob/master/doc/naming.md, which allows unix, dns, and unix-abstract schemes.  It defaults to dns as described in the grpc naming doc.  Currently, we do not support TLS, but we could add that in the future.

I updated the description to include those details.

Can you point me to where the docs for external config types is?  I didn't see any other docs in k/k for other apiserver config types

Reply:
I just meant the serialized file - staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go

That's typically what people look at since that matches the config file they're writing (e.g. https://pkg.go.dev/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1#EgressSelectorConfiguration)

Reply:
> It must adhere to https://github.com/grpc/grpc/blob/master/doc/naming.md, which allows unix, dns, and unix-abstract schemes. It defaults to dns as described in the grpc naming doc. Currently, we do not support TLS, but we could add that in the future.

should that be checked reflected in `validateEndpoint`? if something that parses as a URL but is malformed as a trace exporter endpoint is specified, does kube-apiserver fail to start or does it run in a degraded way?

Reply:
Validation already happens when the connection tries to be established from what I can tell. But we could provide a more clear error log msg with a `validateEndpoint` helper function. And in either case we should not fail or be degraded if tracing is misconfigured, agreed 👍

Reply:
If the endpoint isn't valid (or not reachable), opentelemetry will periodically try to reconnect.  While disconnected, it will buffer (by default up to ~2k) spans, and then drop spans after the buffer is full.  Other than using memory for buffering spans, there won't be any degraded behavior since sending spans is async from the serving path.

I added some additional validation on the scheme.  Mostly, this will catch users that try and specify http(s) as the scheme.

Reply:
sounds good, those seem like likely errors it's good to catch early

Reply:
Yeah I agree with @liggitt, let's catch this at boot rather than at runtime.

Reply:
or CRD load time or whatever that point is..